docs/docs/explainers/explainer-writing-noir.md

```diff
@@ -0,0 +1,178 @@
+
+But there is a twist, the cost of writing to disk is amplified, and is a few orders of magnitude larger. Namely: 20k for allocating and writing a new value, or 5k for writing an existing value.
+So whilst writing efficient code for regular computers and the EVM is mostly the same, there are some key differences like this more immediate compensation of disk writes.
+
+In scenarios where extremely low gas costs are required for an application to be viable/competitive, developers get into what is colloquially known as: "*gas golfing*". Finding the lowest execution cost to achieve a specific task.
+
+### Coding for circuits - a paradigm shift
+
+In zero knowledge cryptography, code is compiled to arithmetic gates called "circuits", and gate count is the significant cost. Depending on the backend this is linearly proportionate to proving time, and so from a product point this should be kept as low as possible.
+
+Whilst writing efficient code for web apps and solidity has a few key differences, writing efficient circuits have a different set of considerations. It is a bit of a paradigm shift, like writing code for GPUs for the first time...
+
+Eg drawing a circle at (0, 0) of radius `r`:
+- For a single CPU thread,
+```
+for theta in 0..2*pi {
+  let x = r * cos(theta);
+  let y = r * sin(theta);
+  draw(x, y);
+} // note: would do 0 - pi/2 and draw +ve/-ve x and y.
+```
+
+- For GPUs (simultaneous parallel calls with x, y across image),
+```
+if (x^2 + y^2 = r^2) {
+  draw(x, y);
+}
+```
+
+([Related](https://www.youtube.com/watch?v=-P28LKWTzrI))
+
+Whilst this CPU -> GPU does not translate to circuits exactly, it is intended to exemplify the difference in intuition when coding for different machine capabilities/constraints.
+
+### Context Takeaway
+
+For those coming from a primarily web app background, this article will explain what you need to consider when writing circuits. Furthermore, for those experienced writing efficient machine code, prepare to shift what you think is efficient 😬
+
+## Code re-use
+
+For some applications using Noir, existing code might be a convenient starting point to then proceed to optimize the gate count of.
+
+:::note
+Many valuable functions and algorithms have been written in more established languages (C/C++), and converted to modern ones (like Rust).
+:::
+
+Fortunately for Noir developers, when needing a particular function a rust implementation can be readily compiled into Noir with some key changes. While the compiler does a decent amount of optimizations, it won't be able to change code that has been optimized for clock-cycles into code optimized for arithmetic gates.
+
+A few things to do when converting Rust code to Noir:
+- `println!` is not a macro, use `println` function (same for `assert_eq`)
+- No early `return` in function. Use constrain via assertion instead
+- No reference `&` operator. Remove, will be value copy
```

Comment:
I am unsure what this line actually means. Can you elaborate a bit?

Reply:
Noir does not have the reference operator like in Rust etc, so it can only pass by value (ie making a copy). I'll clarify this line.